devlog/2024-02-09-Intro-to-optimizing-packages-on-solus.md

```diff
@@ -0,0 +1,488 @@
+
+```
+$ dwebp -h
+Usage: dwebp in_file [options] [-o out_file]
+
+Decodes the WebP image file to PNG format [Default].
+Note: Animated WebP files are not supported.
+$ cwebp -h
+Usage:
+
+   cwebp [options] -q quality input.png -o output.webp
+```
+
+Awesome, these binaries do exactly what we need to benchmark `libwebp`, but, we are also indirectly testing `libpng` as well for this benchmark, we'll have to keep an eye out for that.
+
+One extra step we have to do is ensure these binaries are actually linking against their own library, as upstream developers can have a habit of making sure their binaries don't link against their own libraries and end up being self-contained. Run `ldd` to verify.
+
+```
+$ ldd /usr/bin/dwebp
+	linux-vdso.so.1 (0x00007ffed8733000)
+	libwebpdemux.so.2 => /usr/lib/libwebpdemux.so.2.0.14 (0x00007f7473bb4000)
+	libwebp.so.7 => /usr/lib/libwebp.so.7.1.8 (0x00007f7473ae2000)
+	libpng16.so.16 => /usr/lib/libpng16.so.16 (0x00007f7473aa6000)
+	libc.so.6 => /usr/lib/glibc-hwcaps/x86-64-v3/libc.so.6 (0x00007f74738a9000)
+	libsharpyuv.so.0 => /usr/lib/libsharpyuv.so.0.0.1 (0x00007f747389e000)
+	libm.so.6 => /usr/lib/glibc-hwcaps/x86-64-v3/libm.so.6 (0x00007f74737b8000)
+	libz.so.1 => /usr/lib/libz.so.1.3.0 (0x00007f7473200000)
+	/usr/lib64/ld-linux-x86-64.so.2 (0x00007f7473bea000)
+```
+
+Awesome in this case both `dwebp` and `cwebp` link against `libwebp.so` so we can be confident that any performance improvements will be applicable to all packages in the repository linking against `libwebp`.
+
+Let's grab a couple of `.webp` files from [here](https://developers.google.com/speed/webp/gallery1) to test with. We'll just use the largest image size available in this case to reduce noise as much as possible when running benchmarks as well as allow any potential optimizations to stetch their legs a little more.
+
+Now having done the prep work, lets actually benchmark the damn thing using `hyperfine` for the time being.
+
+```
+$ hyperfine "dwebp ~/3.webp -o /dev/null"
+Benchmark 1: dwebp ~/3.webp -o /dev/null
+  Time (mean ± σ):     202.2 ms ±   0.3 ms    [User: 198.9 ms, System: 3.0 ms]
+  Range (min … max):   201.8 ms … 202.7 ms    14 runs
+$ hyperfine "cwebp ~/PNG_Test.png -o /dev/null"
+Benchmark 1: cwebp ~/PNG_Test.png -o /dev/null
+  Time (mean ± σ):      1.423 s ±  0.009 s    [User: 1.346 s, System: 0.076 s]
+  Range (min … max):    1.410 s …  1.435 s    10 runs
+```
+
+There, we have our a basic baseline for encode and decode performance. We mostly care about decode performance here but improved encoding performance would also not go amiss.
+
+## Let's Start Obvious
+
+Let's start basic, enabling the `speed` optmization which basically builds with `-O3` instead of `-O2` as well as any other flags that are deemed to be worthy to be part of the `speed` group. As well as, the `lto` optmization which builds with link time optimization allowing for interprocedural optimizations to take place.
```

Comment:
```suggestion
Let's start basic, enabling the `speed` optimization which basically builds with `-O3` instead of `-O2` as well as any other flags that are deemed to be worthy to be part of the `speed` group. As well as, the `lto` optimization which builds with link time optimization allowing for inter-procedural optimizations to take place.
```